### ISSUE TITLE
Module extensions using readdir(watch = "yes") are evaluated many times

### ISSUE THREAD

**Issue by calliecameron** [bazelbuild/bazel#23989]
### Description of the bug:

Module extensions using `readdir(watch = "yes")` to watch dir listings in the root workspace have their evaulation restarted each time `readdir` is called. In the example below, we run a command to find dirs, then call readdir on each of them, and you can see from the trace that evaluation gets restarted once for each dir, and has to run the command again each time.

The problem here is that if the command being run is slow and there are many dirs, this can take a long time, and gets slower the more dirs there are. I'd expect to only have to run the command once, and have the first evaluation complete.

### Which category does this issue belong to?

Core

### What's the simplest, easiest way to reproduce this bug? Please provide a minimal example if possible.

```text
$ ls -a
.bazelversion  BUILD  MODULE.bazel  example.bzl

$ mkdir foo bar

$ cat .bazelversion
7.3.2

$ cat BUILD

$ cat MODULE.bazel
module(
    name = "test_module",
    version = "0.0.0",
)

example = use_extension("//:example.bzl", "example")

$ cat example.bzl
def _example_impl(module_ctx):
    print("MODULE EXTENSION START")
    root = module_ctx.path(Label("@@//:WORKSPACE")).dirname

    result = module_ctx.execute(["find", str(root), "-type", "d"])

    if result.return_code != 0:
        fail("Error!")

    for dir in result.stdout.strip("\n").split("\n"):
        print("WATCHING " + dir)
        root.get_child(dir).readdir(watch = "yes")
        print("WATCHED " + dir)

    print("MODULE EXTENSION END")

    return module_ctx.extension_metadata(reproducible = True)

example = module_extension(
    implementation = _example_impl,
)
```

Running `bazel mod deps` here prints:

```text
DEBUG: /home/callie/bazel-test/example.bzl:2:10: MODULE EXTENSION START
DEBUG: /home/callie/bazel-test/example.bzl:2:10: MODULE EXTENSION START
DEBUG: /home/callie/bazel-test/example.bzl:11:14: WATCHING /home/callie/bazel-test
DEBUG: /home/callie/bazel-test/example.bzl:2:10: MODULE EXTENSION START
DEBUG: /home/callie/bazel-test/example.bzl:11:14: WATCHING /home/callie/bazel-test
DEBUG: /home/callie/bazel-test/example.bzl:13:14: WATCHED /home/callie/bazel-test
DEBUG: /home/callie/bazel-test/example.bzl:11:14: WATCHING /home/callie/bazel-test/bar
DEBUG: /home/callie/bazel-test/example.bzl:2:10: MODULE EXTENSION START
DEBUG: /home/callie/bazel-test/example.bzl:11:14: WATCHING /home/callie/bazel-test
DEBUG: /home/callie/bazel-test/example.bzl:13:14: WATCHED /home/callie/bazel-test
DEBUG: /home/callie/bazel-test/example.bzl:11:14: WATCHING /home/callie/bazel-test/bar
DEBUG: /home/callie/bazel-test/example.bzl:13:14: WATCHED /home/callie/bazel-test/bar
DEBUG: /home/callie/bazel-test/example.bzl:11:14: WATCHING /home/callie/bazel-test/foo
DEBUG: /home/callie/bazel-test/example.bzl:2:10: MODULE EXTENSION START
DEBUG: /home/callie/bazel-test/example.bzl:11:14: WATCHING /home/callie/bazel-test
DEBUG: /home/callie/bazel-test/example.bzl:13:14: WATCHED /home/callie/bazel-test
DEBUG: /home/callie/bazel-test/example.bzl:11:14: WATCHING /home/callie/bazel-test/bar
DEBUG: /home/callie/bazel-test/example.bzl:13:14: WATCHED /home/callie/bazel-test/bar
DEBUG: /home/callie/bazel-test/example.bzl:11:14: WATCHING /home/callie/bazel-test/foo
DEBUG: /home/callie/bazel-test/example.bzl:13:14: WATCHED /home/callie/bazel-test/foo
DEBUG: /home/callie/bazel-test/example.bzl:15:10: MODULE EXTENSION END
```

### Which operating system are you running Bazel on?

Ubuntu 22.04

### What is the output of `bazel info release`?

release 7.3.2

### If `bazel info release` returns `development version` or `(@non-git)`, tell us how you built Bazel.

N/A

### What's the output of `git remote get-url origin; git rev-parse HEAD` ?

```text
N/A
```

**Comment by fmeum** [comment]
This is fixed by https://github.com/bazelbuild/bazel/commit/bf27b7dc584d20fab0100103b1419ae9ba66564f, which gets rid of restarts entirely. This commit will be in 8.0.0 and 7.4.0.

**Comment by calliecameron** [comment]
Confirmed, this is fixed in 7.4.0rc2 and 8.0.0rc1. Thanks!
